File: srcs/server/agents/builtin/llm_anthropic.go
package builtin

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// AnthropicClient implements LLMClient for Anthropic's Claude API.
type AnthropicClient struct {
	APIKey string
	Client *http.Client
}

func NewAnthropicClient(apiKey string) *AnthropicClient {
	return &AnthropicClient{
		APIKey: apiKey,
		Client: &http.Client{Timeout: 2 * time.Minute},
	}
}

func (c *AnthropicClient) Chat(ctx context.Context, req ChatRequest) (ChatResponse, error) {
	// Anthropic Messages API expects system prompt as a separate field, not in messages
	// For simplicity, we assume req structure aligns with a translation layer, or we map it here.

	// Map our ChatRequest to Anthropic's payload
	type antMessage struct {
		Role    string `json:"role"`
		Content string `json:"content"`
	}

	var messages []antMessage
	for _, m := range req.Messages {
		if m.Role == RoleSystem {
			continue // Handled separately
		}

		role := string(m.Role)
		if role == string(RoleTool) {
			role = "user" // Tool results are often sent as user in some simplified layers
		}

		messages = append(messages, antMessage{
			Role:    role,
			Content: m.Content,
		})
	}

	payload := map[string]interface{}{
		"model":      req.Model,
		"max_tokens": req.MaxTokens,
		"system":     req.System,
		"messages":   messages,
	}

	body, _ := json.Marshal(payload)

	httpReq, err := http.NewRequestWithContext(ctx, "POST", "https://api.anthropic.com/v1/messages", bytes.NewReader(body))
	if err != nil {
		return ChatResponse{}, err
	}

	httpReq.Header.Set("x-api-key", c.APIKey)
	httpReq.Header.Set("anthropic-version", "2023-06-01")
	httpReq.Header.Set("content-type", "application/json")

	resp, err := c.Client.Do(httpReq)
	if err != nil {
		return ChatResponse{}, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return ChatResponse{}, fmt.Errorf("anthropic api error: %s", resp.Status)
	}

	var result struct {
		Content []struct {
			Text string `json:"text"`
		} `json:"content"`
		Usage struct {
			InputTokens  int `json:"input_tokens"`
			OutputTokens int `json:"output_tokens"`
		} `json:"usage"`
		StopReason string `json:"stop_reason"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return ChatResponse{}, err
	}

	content := ""
	if len(result.Content) > 0 {
		content = result.Content[0].Text
	}

	return ChatResponse{
		Message: Message{
			Role:    RoleAssistant,
			Content: content,
		},
		Usage: Usage{
			InputTokens:  result.Usage.InputTokens,
			OutputTokens: result.Usage.OutputTokens,
		},
		StopReason: result.StopReason,
	}, nil
}

File: srcs/server/agents/builtin/llm_anthropic.cc
#include "srcs/server/agents/builtin/llm_anthropic.h"

#include <utility>

#include "srcs/server/agents/builtin/http_client.h"
#include "srcs/server/agents/builtin/llm_parsing.h"
#include "absl/status/status.h"
#include "absl/strings/str_cat.h"
#include "nlohmann/json.hpp"

namespace ohc::agent {

AnthropicClient::AnthropicClient(std::string api_key,
                                 std::string api_version,
                                 std::string endpoint)
    : api_key_(std::move(api_key)),
      api_version_(std::move(api_version)),
      endpoint_(std::move(endpoint)) {}

absl::StatusOr<ChatResponse> AnthropicClient::Chat(const ChatRequest& req) {
  // ---- Build messages array (Anthropic format) ----------------------------
  // Anthropic requires alternating user/assistant turns and treats tool calls
  // as content blocks.  For this simplified mapping we flatten everything into
  // text content so it works with the current tool schema.
  nlohmann::json messages = nlohmann::json::array();

  for (const auto& m : req.messages) {
    if (m.role == Role::kSystem) continue;  // Handled via top-level "system".

    std::string role(RoleToStringView(m.role));
    // Tool results travel back as a "user" role in Anthropic's format.
    if (m.role == Role::kTool) role = "user";

    nlohmann::json content = nlohmann::json::array();

    // Textual content block.
    if (!m.content.empty()) {
      content.push_back({{"type", "text"}, {"text", m.content}});
    }

    // Tool-use result blocks.
    for (const auto& tr : m.tool_results) {
      content.push_back({
          {"type",        "tool_result"},
          {"tool_use_id", tr.tool_call_id},
          {"content",     tr.content.empty() ? tr.error : tr.content},
      });
    }

    // Tool-use request blocks (from previous assistant turn).
    for (const auto& tc : m.tool_calls) {
      content.push_back({
          {"type",  "tool_use"},
          {"id",    tc.id},
          {"name",  tc.name},
          {"input", tc.arguments},
      });
    }

    if (!content.empty()) {
      messages.push_back({{"role", role}, {"content", std::move(content)}});
    }
  }

  // ---- Build payload -------------------------------------------------------
  nlohmann::json payload;
  payload["model"]      = req.model;
  payload["max_tokens"] = req.max_tokens > 0 ? req.max_tokens : 2048;
  payload["messages"]   = std::move(messages);
  if (!req.system.empty()) payload["system"] = req.system;

  if (!req.tool_definitions.empty()) {
    nlohmann::json tools = nlohmann::json::array();
    for (const auto& td : req.tool_definitions) {
      tools.push_back({
          {"name",         td.name},
          {"description",  td.description},
          {"input_schema", td.parameters},
      });
    }
    payload["tools"] = std::move(tools);
  }

  // ---- HTTP POST -----------------------------------------------------------
  HttpRequest http_req;
  http_req.url    = endpoint_;
  http_req.method = "POST";
  http_req.body   = payload.dump();
  http_req.headers = {
      {"x-api-key",          api_key_},
      {"anthropic-version",  api_version_},
      {"content-type",       "application/json"},
  };

  auto resp_or = HttpDo(http_req);
  if (!resp_or.ok()) return resp_or.status();
  if (resp_or->status_code != 200) {
    return absl::InternalError(absl::StrCat(
        "Anthropic API HTTP ", resp_or->status_code, ": ", resp_or->body));
  }

  // ---- Parse response ------------------------------------------------------
  return ParseAnthropicChatResponse(resp_or->body);
}

}  // namespace ohc::agent
